File: src/MagnusAddon/ui/web/vocab/vocab_sentences_vocab_sentence_view_model.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from sysutils.typed import non_optional

if TYPE_CHECKING:
    from note.sentences.parsed_word import ParsedMatch
    from note.sentences.parsing_result import ParsingResult
    from note.sentences.sentencenote import SentenceNote
    from note.vocabulary.vocabnote import VocabNote

class VocabSentenceMatchViewModel(Slots):
    def __init__(self, match: ParsedMatch, sentence_view_model: VocabSentenceViewModel) -> None:
        self.match: ParsedMatch = match
        self.sentence_view_model: VocabSentenceViewModel = sentence_view_model

    @property
    def is_displayed(self) -> bool: return self.match.is_displayed
    @property
    def start_index(self) -> int: return self.match.start_index
    @property
    def end_index(self) -> int: return self.match.end_index
    def is_primary_form_of(self, vocab: VocabNote) -> bool: return self.match.parsed_form == vocab.question.without_noise_characters
    @property
    def shaded_by(self) -> ParsedMatch | None:
        shading = [match for match in reversed(self.sentence_view_model.result.parsed_words)
                   if match.is_displayed
                   and match.start_index <= self.start_index <= match.end_index]

        return shading[0] if shading else None

    @property
    def yields_to(self) -> ParsedMatch | None:
        yields_to = [match for match in self.sentence_view_model.result.parsed_words
                     if match.is_displayed
                     and self.start_index < match.start_index <= self.end_index < match.end_index]

        return yields_to[0] if yields_to else None

    @override
    def __repr__(self) -> str: return self.match.__repr__()

class VocabSentenceViewModel(Slots):
    def __init__(self, _vocab_note: VocabNote, sentence_note: SentenceNote) -> None:
        self.vocab: VocabNote = _vocab_note
        self.sentence: SentenceNote = sentence_note
        self.result: ParsingResult = sentence_note.parsing_result.get()
        self.matches: list[VocabSentenceMatchViewModel] = [VocabSentenceMatchViewModel(match, self) for match in self.result.parsed_words if match.vocab_id == _vocab_note.get_id()]
        self.displayed_matches: list[VocabSentenceMatchViewModel] = [match for match in self.matches if match.is_displayed]
        self.highlighted_sentences: set[SentenceNote] = set(_vocab_note.sentences.user_highlighted())
        self.shaded_matches: list[VocabSentenceMatchViewModel] = [match for match in self.matches if not match.is_displayed]
        self.matched_vocab_ids: set[int] = {match.vocab_id for match in self.result.parsed_words}

    @property
    def primary_match(self) -> VocabSentenceMatchViewModel: return self.displayed_matches[0] if self.displayed_matches else self.matches[0]

    def format_sentence(self) -> str:
        result = self.result

        match = self.primary_match
        is_primary = "primary" if match.is_primary_form_of(self.vocab) else "secondary"
        match_class = f"{is_primary}FormMatch"

        if match.is_displayed:
            head = result.sentence[:match.start_index]
            match_range = result.sentence[match.start_index:match.end_index]
            tail = result.sentence[match.end_index:]
            return f"""{head}<span class="vocabInContext {match_class}">{match_range}</span>{tail}"""
        else:  # noqa: RET505
            covering_match = non_optional(match.shaded_by or match.yields_to)
            covering_match_class = "compound" if covering_match.vocab_id in self.vocab.related_notes.in_compound_ids else "shadingMatch"
            head = result.sentence[:min(covering_match.start_index, match.start_index)]
            shading_pre_range = result.sentence[covering_match.start_index:match.start_index]
            match_range = result.sentence[match.start_index:match.end_index]
            shading_post_range = result.sentence[match.end_index:covering_match.end_index]
            tail = result.sentence[max(covering_match.end_index, match.end_index):]
            return "".join([head,
                            f"""<span class="vocabInContext {match_class} {covering_match_class}">{shading_pre_range}</span>""",
                            f"""<span class="vocabInContext {match_class}">{match_range}</span>""",
                            f"""<span class="vocabInContext {match_class} {covering_match_class}">{shading_post_range}</span>""",
                            tail])

    def is_highlighted(self) -> bool: return self.sentence in self.highlighted_sentences

    @property
    def vocab_is_displayed(self) -> bool: return any(self.displayed_matches)

    def sentence_classes(self) -> str:
        classes = ""
        if self.sentence in self.highlighted_sentences: classes += "highlighted "
        classes += " ".join(self.sentence.get_meta_tags())
        return classes

    def contains_primary_form(self) -> bool:
        return any(match for match in self.displayed_matches if match.is_primary_form_of(self.vocab))

File: src/src_dotnet/JAStudio.Core/UI/Web/Vocab/VocabSentencesVocabSentenceViewModel.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note;
using JAStudio.Core.Note.Sentences;
using JAStudio.Core.Note.Vocabulary;

namespace JAStudio.Core.UI.Web.Vocab;

public class VocabSentenceMatchViewModel
{
   public ParsedMatch Match { get; }
   public VocabSentenceViewModel SentenceViewModel { get; }

   public VocabSentenceMatchViewModel(ParsedMatch match, VocabSentenceViewModel sentenceViewModel)
   {
      Match = match;
      SentenceViewModel = sentenceViewModel;
   }

   public bool IsDisplayed => Match.IsDisplayed;
   public int StartIndex => Match.StartIndex;
   public int EndIndex => Match.EndIndex;

   public bool IsPrimaryFormOf(VocabNote vocab) => Match.ParsedForm == vocab.Question.WithoutNoiseCharacters;

   public ParsedMatch? ShadedBy
   {
      get
      {
         var allParsedWords = SentenceViewModel.Result.ParsedWords.ToList();
         allParsedWords.Reverse();
         var shading = allParsedWords
                      .Where(match => match.IsDisplayed && match.StartIndex <= StartIndex && StartIndex <= match.EndIndex)
                      .ToList();

         return shading.Any() ? shading[0] : null;
      }
   }

   public ParsedMatch? YieldsTo
   {
      get
      {
         var yieldsTo = SentenceViewModel.Result.ParsedWords
                                         .Where(match => match.IsDisplayed &&
                                                         StartIndex < match.StartIndex &&
                                                         match.StartIndex <= EndIndex &&
                                                         EndIndex < match.EndIndex)
                                         .ToList();

         return yieldsTo.Any() ? yieldsTo[0] : null;
      }
   }

   public override string ToString() => Match.ToString() ?? string.Empty;
}

public class VocabSentenceViewModel
{
   public VocabNote Vocab { get; }
   public SentenceNote Sentence { get; }
   public ParsingResult Result { get; }
   public List<VocabSentenceMatchViewModel> Matches { get; }
   public List<VocabSentenceMatchViewModel> DisplayedMatches { get; }
   public HashSet<SentenceNote> HighlightedSentences { get; }
   public List<VocabSentenceMatchViewModel> ShadedMatches { get; }
   public HashSet<NoteId> MatchedVocabIds { get; }

   public VocabSentenceViewModel(VocabNote vocabNote, SentenceNote sentenceNote)
   {
      Vocab = vocabNote;
      Sentence = sentenceNote;
      Result = sentenceNote.GetParsingResult();
      Matches = Result.ParsedWords
                      .Where(match => match.VocabId == vocabNote.GetId())
                      .Select(match => new VocabSentenceMatchViewModel(match, this))
                      .ToList();
      DisplayedMatches = Matches.Where(match => match.IsDisplayed).ToList();
      HighlightedSentences = vocabNote.Sentences.UserHighlighted().ToHashSet();
      ShadedMatches = Matches.Where(match => !match.IsDisplayed).ToList();
      MatchedVocabIds = Result.ParsedWords.Where(match => match.VocabId != null).Select(match => match.VocabId!).ToHashSet();
   }

   public VocabSentenceMatchViewModel PrimaryMatch =>
      DisplayedMatches.Any() ? DisplayedMatches[0] : Matches[0];

   public string FormatSentence()
   {
      var result = Result;
      var match = PrimaryMatch;
      var isPrimary = match.IsPrimaryFormOf(Vocab) ? "primary" : "secondary";
      var matchClass = $"{isPrimary}FormMatch";

      if(match.IsDisplayed)
      {
         var head = result.Sentence.Substring(0, match.StartIndex);
         var matchRange = result.Sentence.Substring(match.StartIndex, match.EndIndex - match.StartIndex);
         var tail = result.Sentence.Substring(match.EndIndex);
         return $"{head}<span class=\"vocabInContext {matchClass}\">{matchRange}</span>{tail}";
      } else
      {
         var shadingMatch = match.ShadedBy;

         var shadingStartIndex = shadingMatch?.StartIndex ?? match.StartIndex;
         var shadingEndIndex = shadingMatch?.EndIndex ?? match.EndIndex;

         var preCoveringMatchClass = shadingMatch != null && shadingMatch.VocabId != null && Vocab.RelatedNotes.InCompoundIds.Contains(shadingMatch.VocabId)
                                        ? "compound"
                                        : "shadingMatch";
         var postCoveringMatchClass = preCoveringMatchClass;

         var yieldsTo = match.YieldsTo;
         if(yieldsTo != null)
         {
            shadingEndIndex = yieldsTo.EndIndex;
            postCoveringMatchClass = "yieldsTo";
         }

         var sentenceText = result.Sentence;
         var minStart = System.Math.Min(shadingStartIndex, match.StartIndex);
         var maxEnd = System.Math.Max(shadingEndIndex, match.EndIndex);

         var head = sentenceText.Substring(0, minStart);
         var shadingPreRange = sentenceText.Substring(System.Math.Min(shadingStartIndex, match.StartIndex), System.Math.Max(0, match.StartIndex - shadingStartIndex));
         var matchRange = sentenceText.Substring(match.StartIndex, match.EndIndex - match.StartIndex);
         var shadingPostRange = sentenceText.Substring(match.EndIndex, System.Math.Max(0, shadingEndIndex - match.EndIndex));
         var tail = sentenceText.Substring(maxEnd);

         return string.Concat(
            head,
            $"<span class=\"vocabInContext {matchClass} {preCoveringMatchClass}\">{shadingPreRange}</span>",
            $"<span class=\"vocabInContext {matchClass}\">{matchRange}</span>",
            $"<span class=\"vocabInContext {matchClass} {postCoveringMatchClass}\">{shadingPostRange}</span>",
            tail);
      }
   }

   public bool IsHighlighted() => HighlightedSentences.Contains(Sentence);

   public bool VocabIsDisplayed => DisplayedMatches.Any();

   public string SentenceClasses()
   {
      var classes = "";
      if(HighlightedSentences.Contains(Sentence)) classes += "highlighted ";
      classes += string.Join(" ", Sentence.GetMetaTags());
      return classes;
   }

   public bool ContainsPrimaryForm()
   {
      return DisplayedMatches.Any(match => match.IsPrimaryFormOf(Vocab));
   }
}
